### ISSUE TITLE
Request for an open review criteria

### ISSUE THREAD

**Issue by jonocarroll** [cranchange/cran_change.org#8]
Thank you, Miles, for this letter. I believe the time for change is upon us and something formal needs to be presented.

One aspect I think would benefit the relationship between the CRAN team and package authors is an open (public, visible, human-readable) review guide to which package authors can refer to confirm that their package _should_ be accepted. We currently have the [CRAN Repository Policy](https://cran.r-project.org/web/packages/policies.html) and the source of [`R CMD check`](https://github.com/wch/r-source/blob/trunk/src/library/tools/R/check.R) itself but neither of these is a comprehensive summary of what is strictly required, and neither is particularly clear. The [canonical checklist](https://cran.r-project.org/web/packages/submission_checklist.html) falls woefully short of a reliable guide.

Some details of the checks are spread out over other pages on the CRAN website, such as the details regarding [URL checks](https://cran.r-project.org/web/packages/URL_checks.html) (since 301 redirects are a common cause for submission issues).

There are external resources that help meet this need, for example [R-pkgs](https://r-pkgs.org/r-cmd-check.html) has a great guide explaining some of these checks and what they actually mean if they go wrong. This should be something provided by R itself since that's where the checks are defined, but if CRAN is enforcing these, then it would be helpful for them to either limit their conditions to these checks or detail exactly which checks will be performed and how to satisfy those. I also believe the CRAN team would be in a suitable position to have their full suite of checks added to the official R checks so that these are distributed and can be checked by package authors more reliably.

Other resources towards this goal are spread around and occasionally updated, e.g.
* https://github.com/ThinkR-open/prepare-for-cran
* https://github.com/DavisVaughan/extrachecks

One point which comes up frequently enough is that restrictions on using `\donttest{}` are not published (at least anywhere I've seen). Package authors are encouraged to add examples to their code (as they should) but various packages rely on external resources which may fail at random on CRAN's test infrastructure. Package authors, wishing to retain some examples but acknowledging that these may not be robust to missing external resources, use some provided functionality but have had submissions rejected if they use this approach.

Requirements are often at the discretion of any particular reviewer, and it is all too well known that there is one with stronger requirements than others
> I just heard from a friend that the renowned professor complained "It's" must be "It is" in his email to CRAN. 
https://twitter.com/xieyihui/status/508650816001945600?s=20

Rejection of a package because the functionality was [deemed unnecessary](https://twitter.com/HollyKirk/status/1060272895920812032?s=20) by a reviewer makes submission a questionable effort. This is [apparently](https://twitter.com/hadleywickham/status/997493612798656512?s=20) not an isolated case. The [requirement](https://cran.r-project.org/web/packages/policies.html#Preamble) that 
> A package’s contribution has to be non-trivial

is too vague to be enforceable. The criteria should detail what a package must contain in order to be considered for submission.

Providing the actual review criteria (and sticking to them, updating them if necessary) to which package maintainers could align their submissions would strengthen the relationship between the community of package maintainers and the CRAN team. I would love to see such criteria be open to discussion for clarification, extension, and refinement. The moving goalpost in place at the moment will only lead to an abandonment of the service which is of great value to the community (when it works).

**Comment by MilesMcBain** [comment]
Excellent points @jonocarroll! @adamhsparks this seems to articulate the issues you were referring to regarding examples quite well?

**Comment by adamhsparks** [comment]
Yes, thanks, @jonocarroll. That's exactly what I was trying to formulate.

We need a consistent method for handling external resources that may fail in examples and tests. This doesn't mean that the package is broken, but CRAN treats them in this fashion. To my mind, if I request an external resource and it's not available and the package uses `stop()` then it's functioning as intended rather than carrying on like a pork chop and failing downstream. A proper "Fail gracefully" guideline would be most appreciated in these example instances (and maybe tests too but there is just `skip_on_CRAN()` I suppose.

**Comment by adamhsparks** [comment]
Clearly articulating these guidelines would help developers but would also  (presumably) reduce the load that the CRAN team must shoulder.

**Comment by MilesMcBain** [comment]
@adamhsparks I really appreciate your point about `stop()`. If you can't use it for API calls that means everything downstream from that code needs to employ cumbersome status checks, including user code. CRAN shouldn't be enforcing this kind of design.

**Comment by MilesMcBain** [comment]
Thoughts on the above addition? @jonocarroll @adamhsparks 

It probably needs a corresponding reform, around a single authoritative list of acceptance criteria

**Comment by adamhsparks** [comment]
These additions help with tone greatly, @MilesMcBain. I think this is a good direction to take. This illustrates the effects on those of us that submit packages to CRAN rather than making CRAN sound like bad people.

**Comment by JosiahParry** [comment]
This is one of the only issues I have with CRAN at the moment. The requirements for passing CRAN checks differs from what is provided by R CMD Check. For my last submission of a CRAN package I think it took ~5 submissions before I passed all of their checks that were not listed or caught my R CMD Check. It's such an issue that https://github.com/DavisVaughan/extrachecks exists.

**Comment by llrs** [comment]
I recently learned that the automatic CRAN checks are public in the [R-dev-web repository](https://svn.r-project.org/R-dev-web/trunk) [(mirrored in github)](https://github.com/r-devel/r-dev-web). Most of the checks are at the [QA folder](https://github.com/r-devel/r-dev-web/tree/master/CRAN/QA) (see also [this exchange](https://github.com/RConsortium/r-repositories-wg/blob/74761af83a30fd0870c2cd46ba041073fdf08885/Documents/Proposal%20to%20CRAN.md)). 

The [R repository working group](https://github.com/RConsortium/r-repositories-wg) is looking into the checks (look around on that repository issues and minutes for some work in progress, specially https://github.com/RConsortium/r-repositories-wg/issues/17). Hope this helps
